pkg/controller/proxyconfig/controller.go

```diff
@@ -0,0 +1,361 @@
+package proxyconfig
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-network-operator/pkg/controller/statusmanager"
+	"github.com/openshift/cluster-network-operator/pkg/names"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, status *statusmanager.StatusManager) error {
+	reconciler := newReconciler(mgr, status)
+	if reconciler == nil {
+		return fmt.Errorf("failed to create reconciler")
+	}
+
+	return add(mgr, reconciler)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager) reconcile.Reconciler {
+	if err := configv1.Install(mgr.GetScheme()); err != nil {
+		return &ReconcileProxyConfig{}
+	}
+
+	return &ReconcileProxyConfig{client: mgr.GetClient(), scheme: mgr.GetScheme(), status: status}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("proxyconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// We only care about a configmap source with a specific name/namespace,
+	// so filter events before they are provided to the controller event handlers.
+	pred := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.MetaOld.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.MetaOld.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+	}
+
+	// Watch for changes to the trust bundle configmap.
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForObject{}, pred)
```

Comment:
you also need to watch the configmap pointed to by the proxyconfig resource, is that being done somewhere?

Reply:
@bparees `pred` along with `handleConfigMap()` is watching for user-provided configmaps from `openshift-config` ns. The controller merges any a user-provided trust bundle with the system bundle and publishes the merged bundle to `trusted-ca-bundle` configmap in ns `openshift-config-managed`. If `proxy.spec.trustedCA` does not refer to a configmap but `proxy.spec.httpsProxy` is set, then the controller will generate `trsuted-ca-bundle` configmap using the system bundle.

Reply:
>  If proxy.spec.trustedCA does not refer to a configmap but proxy.spec.httpsProxy is set, then the controller will generate trsuted-ca-bundle configmap using the system bundle.

i think it should always generated the bundle regardless of whether httpsProxy is set.  Otherwise the bundle risks going stale if someone unsets httpsProxy.

Reply:
a bundle is being generated regardless of httpsProxy being set. If `trustedCA` is provided, a combined `trustedCA/system` bundle is generated. Otherwise, a bundle is generated containing the system bundle.